/* Copyright 2023 The FedLearner Authors. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

syntax = "proto3";
package fedlearner_webconsole.proto;

enum FilterOp {
  // EQUAL as default
  EQUAL = 0;
  // If it is in a list.
  IN = 1;
  // If it contains a substring.
  CONTAIN = 2;
  GREATER_THAN = 3;
  LESS_THAN = 4;
}

message SimpleExpression {
  message ListValue {
    repeated double number_list = 1;
    repeated string string_list = 2;
  }

  string field = 1;
  FilterOp op = 2;
  oneof value {
    bool bool_value = 3;
    string string_value = 4;
    double number_value = 5;
    ListValue list_value = 6;
  }
}

enum FilterExpressionKind {
  // SimpleExpression
  SIMPLE = 0;
  // Combines sub expressions by AND
  AND = 1;
}

message FilterExpression {
  FilterExpressionKind kind = 1;
  // basic
  SimpleExpression simple_exp = 2;
  // and/or
  repeated FilterExpression exps = 3;
}
